tests/py_wheel/py_wheel_tests.bzl

```diff
@@ -92,8 +97,109 @@ def _test_content_type_from_description_impl(env, target):
 
 _tests.append(_test_content_type_from_description)
 
+def _test_pep440_normalization(env):
+    prefixes = ["v", "  v", " \t\r\nv"]
+    epochs = {
+        "": ["", "0!", "00!"],
+        "1!": ["1!", "001!"],
+        "200!": ["200!", "00200!"],
+    }
+    releases = {
+        "0.1": ["0.1", "0.01"],
+        "2023.7.19": ["2023.7.19", "2023.07.19"],
+    }
+    pres = {
+        "": [""],
+        "a0": ["a", ".a", "-ALPHA0", "_alpha0", ".a0"],
+        "a4": ["alpha4", ".a04"],
+        "b0": ["b", ".b", "-BETA0", "_beta0", ".b0"],
+        "b5": ["beta05", ".b5"],
+        "rc0": ["C", "_c0", "RC", "_rc0", "-preview_0"],
+    }
+    explicit_posts = {
+        "": [""],
+        ".post0": [],
+        ".post1": [".post1", "-r1", "_rev1"],
+    }
+    implicit_posts = [[".post1", "-1"], [".post2", "-2"]]
+    devs = {
+        "": [""],
+        ".dev0": ["dev", "-DEV", "_Dev-0"],
+        ".dev9": ["DEV9", ".dev09", ".dev9"],
+        ".dev{BUILD_TIMESTAMP}": [
+            "-DEV{BUILD_TIMESTAMP}",
+            "_dev_{BUILD_TIMESTAMP}",
+        ],
+    }
+    locals = {
+        "": [""],
+        "+ubuntu.7": ["+Ubuntu_7", "+ubuntu-007"],
+        "+ubuntu.r007": ["+Ubuntu_R007"],
+    }
+    epochs = [
+        [normalized_epoch, input_epoch]
+        for normalized_epoch, input_epochs in epochs.items()
+        for input_epoch in input_epochs
+    ]
+    releases = [
+        [normalized_release, input_release]
+        for normalized_release, input_releases in releases.items()
+        for input_release in input_releases
+    ]
+    pres = [
+        [normalized_pre, input_pre]
+        for normalized_pre, input_pres in pres.items()
+        for input_pre in input_pres
+    ]
+    explicit_posts = [
+        [normalized_post, input_post]
+        for normalized_post, input_posts in explicit_posts.items()
+        for input_post in input_posts
+    ]
+    pres_and_posts = [
+        [normalized_pre + normalized_post, input_pre + input_post]
+        for normalized_pre, input_pre in pres
+        for normalized_post, input_post in explicit_posts
+    ] + [
+        [normalized_pre + normalized_post, input_pre + input_post]
+        for normalized_pre, input_pre in pres
+        for normalized_post, input_post in implicit_posts
+        if input_pre == "" or input_pre[-1].isdigit()
+    ]
+    devs = [
+        [normalized_dev, input_dev]
+        for normalized_dev, input_devs in devs.items()
+        for input_dev in input_devs
+    ]
+    locals = [
+        [normalized_local, input_local]
+        for normalized_local, input_locals in locals.items()
+        for input_local in input_locals
+    ]
+    postfixes = ["", "  ", " \t\r\n"]
+    i = 0
+    for nepoch, iepoch in epochs:
```

Comment:
I agree that we should be able to remove some of the unit tests that overlap if we're smart about it. I'm not sold on testing the individual helper functions (e.g. epoch parsing) and then having a single integration test.

What is our purpose with having tests here? Arguably, one purpose is to find bugs, and then the main consideration is what bugs we expect to find. This is then best informed by what sort of bugs I did find when developing this, and apart from the trivial it was what you might call integration issues between different parts of the parser, as exemplified by the `0bdev` issue, where the pre-release parser was broken in the presence of a dev release.

Another purpose that we shouldn't discount is to facilitate future changes. What future changes do we anticipate? Not changes to the requirements of PEP 440, but, as you point out, we may refactor the code as new techniques on writing good starlark code are developed. A refactoring is helped a lot by a reasonably comprehensive test suite _that you don't have to change_ when you refactor the code. Unit tests that guard the current detailed design aren't useful; tests that guard the unchanging requirements are useful. If we replace the parser with regular expressions, there won't be an epoch parser as a distinct thing. Likewise if we replace the recursive descent parser with a table-based parser generated by a parser generator using a grammar definition. The person refactoring this won't thank the test writer if he has to totally rewrite the tests as well. Therefore, I think all the tests should be testing `normalize_pep440()`, not any helper functions it uses, and if we want tests to be smaller and more specific, we should move towards testing the individual requirements of PEP 440.

Reply:
Thanks for the extensive write up, I agree with the points made here, mostly. I think it is important for someone who is seeing the code for the first time to understand what to change if something starts failing. Tests should be also DAMP (https://enterprisecraftsmanship.com/posts/dry-damp-unit-tests/) because it makes them easier to read and understand.

I am OK with resolving this for now.